src/libPMacc/test/random/2DDistribution.cu

```diff
@@ -0,0 +1,223 @@
+/**
+ * Copyright 2016 Alexander Grund
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "pmacc_types.hpp"
+#include "memory/buffers/HostDeviceBuffer.hpp"
+#include "random/RNGProvider.hpp"
+#include "random/distributions/Uniform.hpp"
+#include "random/methods/Xor.hpp"
+#include "random/methods/XorMin.hpp"
+#include "random/methods/MRG32k3a.hpp"
+#include "random/methods/MRG32k3aMin.hpp"
+#include "dimensions/DataSpace.hpp"
+#include <stdint.h>
+#include <iostream>
+#include <fstream>
+
+typedef PMacc::DataSpace<2> Space2D;
+typedef PMacc::DataSpace<3> Space3D;
+
+template<class T_DataBox, class T_Random>
+__global__ void RandomFiller(T_DataBox box, Space2D boxSize, T_Random rand, uint32_t numSamples)
+{
+    const Space3D ownIdx = Space3D(threadIdx) + Space3D(blockIdx) * Space3D(blockDim);
+    rand.init(ownIdx.shrink<2>());
+    for(uint32_t i=0; i<numSamples; i++)
+    {
+        Space2D idx = rand(boxSize);
+        atomicAdd(&box(idx), 1);
+    }
+}
+
+template<class T_RNGProvider>
+struct GetRandomIdx
+{
+    typedef PMacc::random::distributions::Uniform<float> Distribution;
+    typedef typename T_RNGProvider::GetRandomType<Distribution>::type Random;
+
+    HINLINE GetRandomIdx(): rand(T_RNGProvider::template createRandom<Distribution>())
+    {}
+
+    DINLINE void
+    init(Space2D globalCellIdx)
+    {
+        rand.init(globalCellIdx);
+    }
+
+    DINLINE Space2D
+    operator()(Space2D size)
+    {
+        using PMacc::algorithms::math::float2int_rd;
+        return Space2D(float2int_rd(rand() * size.x()), float2int_rd(rand() * size.y()));
```

Comment:
This assumes that rand() is already fixed returning only values in `[0, 1)`
Try to put in the `if(val == 1.f) return 0.f` fix to test this.